cou/apps/auxiliary.py

```diff
@@ -183,6 +183,92 @@ class RabbitMQServer(AuxiliaryApplication):
 
+    def upgrade_plan_sanity_checks(
+        self, target: OpenStackRelease, units: Optional[list[Unit]]
+    ) -> None:
+        """Run sanity checks before generating upgrade plan.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :param units: Units to generate upgrade plan, defaults to None
+        :type units: Optional[list[Unit]], optional
+        :raises ApplicationError: When application is wrongly configured.
+        :raises HaltUpgradePlanGeneration: When the application halt the upgrade plan generation.
+        :raises MismatchedOpenStackVersions: When the units of the app are running
+                                             different OpenStack versions.
+        """
+        self._check_channel()
+        self._check_application_target(target)
+        self._check_mismatched_versions(units)
+        logger.info(
+            "%s application met all the necessary prerequisites to generate the upgrade plan",
+            self.name,
+        )
+
+    def pre_upgrade_steps(
+        self, target: OpenStackRelease, units: Optional[list[Unit]]
+    ) -> list[PreUpgradeStep]:
+        """Pre Upgrade steps planning.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :param units: Units to generate upgrade plan
+        :type units: Optional[list[Unit]]
+        :return: List of pre upgrade steps.
+        :rtype: list[PreUpgradeStep]
+        """
+        steps = super().pre_upgrade_steps(target, units)
+        if self.config["enable-auto-restarts"].get("value") is False:
+            units_to_run_action = self.units.values() if units is None else units
+            steps += [
+                PreUpgradeStep(
+                    description=(
+                        "Enable auto restarts is disabled, : will"
+                        f" run any deferred events and restart services for unit: '{unit.name}'"
+                    ),
+                    coro=self.model.run_action(
+                        unit_name=unit.name,
+                        action_name="run-deferred-hooks",
+                        raise_on_failure=True,
+                    ),
+                )
+                for unit in units_to_run_action
+            ]
+        return steps
+
+    def post_upgrade_steps(
+        self, target: OpenStackRelease, units: Optional[list[Unit]]
+    ) -> list[PostUpgradeStep]:
+        """Post Upgrade steps planning.
+
+        Wait until the application reaches the idle state and then check the target workload.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :param units: Units to generate post upgrade plan
+        :type units: Optional[list[Unit]]
+        :return: List of post upgrade steps.
+        :rtype: list[PostUpgradeStep]
+        """
+        steps = super().post_upgrade_steps(target, units)
```

Comment:
**question:** 

Should we wait until all the unit become idle here?

Reply:
Like this we will wait, since it's part of post_upgrade_steps. However, we should run deferred hooks before this and that means, we should do:
- wait for application
- run deferred hooks
- run normal pos-upgrade steps

Reply:
`super().post_upgrade_steps()` will wait for idle

Reply:
Yes, but you need to wiat first or at least after you run the action.